pymeasure/display/widgets/fileinput_widget.py

```diff
@@ -0,0 +1,153 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2023 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+from ..Qt import QtCore, QtWidgets
+from .filename_widget import FilenameLineEdit
+from .directory_widget import DirectoryLineEdit
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class FileInputWidget(QtWidgets.QWidget):
+    _extensions = ["csv", "txt"]
+
+    def __init__(self, filename_input=True, directory_input=True, parent=None):
+        super().__init__(parent)
+
+        self._setup_ui(filename_input, directory_input)
+        self._layout()
```

Comment:
Yeah, I understand your line of thought, especially with the two `hasattr` checks that were in there.
The idea of this separation is that you define the widgets (that you need/can access later on in the class) in the `setup_ui` method, and have the `layout` method (which can sometimes be quite full and cluttered) to handle where the widgets are placed.

SInce I've removed the `hasattr` check, i'm inclined to keep the structure like this.

Reply:
> SInce I've removed the hasattr check, i'm inclined to keep the structure like this.

It's fine with me to keep the structure, which makes sense to me.

Only the "hasattr" checks made it seem better to unite the methods.